keps/sig-storage/2485-read-write-once-pod-pv-access-mode/README.md

```diff
@@ -0,0 +1,1024 @@
+components? What are the guarantees? Make sure this is in the test plan.
+
+Consider the following in developing a version skew strategy for this
+enhancement:
+- Does this enhancement involve coordinating behavior in the control plane and
+  in the kubelet? How does an n-2 kubelet without this feature available behave
+  when this feature is used?
+- Will any other components on the node change? For example, changes to CSI,
+  CRI or CNI may require updating that component before the kubelet.
+-->
+
+
+#### API Server Version N / Scheduler Version N / Kubelet Version N-1 or N-2
+
+When starting two pods with both using the same PVC with ReadWriteOncePod, one pod
+will successfully start, but the other will not be scheduled due to the
+ReadWriteOncePod access mode conflict.
+
+When starting the same two pods but also setting `pod.spec.nodeName` to the same
+node, kubelet will not enforce the access mode and will proceed with starting
+both pods.
+
+For older kubelets, [ReadWriteOncePod will map to access mode `UNKNOWN`]. How this
+access mode is used will vary across CSI drivers.
+
+[ReadWriteOncePod will map to access mode `UNKNOWN`]: https://github.com/kubernetes/kubernetes/blob/v1.21.0/pkg/volume/csi/csi_client.go#L512
+
+#### API Server Version N / Scheduler Version N-1 / Kubelet Version N
+
+When creating a pod using ReadWriteOncePod, the scheduler will not enforce this
+access mode during scheduling. It will be possible for two pods using the same
+PVC with this access mode to be assigned the same node, however the access mode
+will be enforced by kubelet during mount.
+
+#### API Understands ReadWriteOncePod, CSI Sidecars Do Not
+
+Both the the [CSI attacher] and the [CSI resizer] will error if they do not
+understand ReadWriteOncePod and this access mode is used on a PV.
+
+The CSI provisioner will [map ReadWriteOncePod to a nil access mode]. How this
+access mode is used will vary across CSI drivers.
+
+[CSI attacher]: https://github.com/kubernetes-csi/external-attacher/blob/v3.2.0/pkg/controller/util.go#L196-L197
+[CSI resizer]: https://github.com/kubernetes-csi/external-resizer/blob/v1.2.0/pkg/resizer/csi_resizer.go#L237-L238
+[map ReadWriteOncePod to a nil access mode]: https://github.com/kubernetes-csi/external-provisioner/blob/v2.2.0/pkg/controller/controller.go#L468-L469
+
+#### CSI Sidecars Understand `SINGLE_NODE_{SINGLE,MULTI}_WRITER`, CSI Drivers Do Not
+
+How CSI drivers treat unknown access modes depends per driver.
+
+#### CSI Controller Service Understands `SINGLE_NODE_{SINGLE,MULTI}_WRITER`, CSI Node Service Does Not
+
+If the CSI driver running the controller service understands the new access
+modes, then volumes will be provisioned and attached using these access modes
+(if ReadWriteOncePod or ReadWriteOnce are used). If the CSI driver running the
+node service does not understand these access modes, the behavior will depend on
+the CSI driver and how it treats unknown access modes.
```

Comment:
Maybe we can say here the recommendation is to upgrade both controller and node versions together? Or at least upgrade the node version first?

Reply:
Done